lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -1605,18 +1628,27 @@ struct ConvertLayoutOpConversion
       // TODO: This is actually redundant index calculation, we should
       //       consider of caching the index calculation result in case
       //       of performance issue observed.
-      // for (unsigned elemId = linearCTAId * accumSizePerThread;
-      //      elemId < (linearCTAId + 1) * accumSizePerThread; elemId += vec) {
       for (unsigned elemId = 0; elemId < accumSizePerThread; elemId += vec) {
-        auto multiDimElemId =
-            getMultiDimIndex<unsigned>(elemId, layout.getSizePerThread());
         SmallVector<Value> multiDimOffset(rank);
-        for (unsigned d = 0; d < rank; ++d) {
-          multiDimOffset[d] = add(
-              multiDimOffsetFirstElem[d],
-              createIndexAttrConstant(rewriter, loc, llvmIndexTy,
-                                      multiDimCTAInRepId[d] * shapePerCTA[d] +
-                                          multiDimElemId[d]));
+        if (blockedLayout) {
+          SmallVector<unsigned> multiDimElemId = getMultiDimIndex<unsigned>(
+              elemId, blockedLayout.getSizePerThread());
+          for (unsigned d = 0; d < rank; ++d) {
+            multiDimOffset[d] = rewriter.create<LLVM::AddOp>(
+                loc, multiDimOffsetFirstElem[d],
+                createIndexAttrConstant(rewriter, loc, llvmIndexTy,
+                                        multiDimCTAInRepId[d] * shapePerCTA[d] +
+                                            multiDimElemId[d]));
+          }
+        } else if (mmaLayout) {
+          assert(rank == 2);
```

Comment:
I see lots of assert usage in current code. Is here a real assert, or we are using assert for error handling？

Reply:
it's a real assert, for sure.